File: benchmarks/references/ground-truth/R/tc-015-km-curve.R
#!/usr/bin/env Rscript
# TC-015: Kaplan-Meier Curve with Risk Table
# Generates KM survival curve coordinates and risk table counts
# Output: JSON with curve coordinates, CI, and risk set at specified time points

# --- Dependencies: survival, jsonlite, dplyr ---

args <- commandArgs(trailingOnly = TRUE)
seed <- 42
n <- 200
output <- ""
data_csv <- ""

i <- 1
while (i <= length(args)) {
  if (args[i] == "--seed") { seed <- as.integer(args[i+1]); i <- i + 2 }
  else if (args[i] == "--n") { n <- as.integer(args[i+1]); i <- i + 2 }
  else if (args[i] == "--output") { output <- args[i+1]; i <- i + 2 }
  else if (args[i] == "--data") { data_csv <- args[i+1]; i <- i + 2 }
  else { i <- i + 1 }
}

cat(sprintf("[%s] TC-015 KM Curve + Risk Table | seed=%d n=%d\n", Sys.time(), seed, n))

# --- Data generation (if no shared CSV) ---
generate_adtte <- function(n, seed) {
  set.seed(seed)
  trt <- rep(c(1, 0), each = n %/% 2)
  # Experimental arm: longer median survival (~12 months)
  shape_e <- 1.3; scale_e <- 14
  shape_c <- 1.1; scale_c <- 8
  ttev <- ifelse(trt == 1,
    rgamma(n, shape = shape_e, scale = scale_e),
    rgamma(n, shape = shape_c, scale = scale_c))
  # Censoring
  cens <- runif(n, min = 6, max = 30)
  ev <- as.integer(ttev <= cens)
  ttev <- pmin(ttev, cens)
  # Strata
  sex <- sample(c("F", "M"), n, replace = TRUE)
  ecog <- sample(c(0, 1), n, replace = TRUE, prob = c(0.6, 0.4))
  data.frame(
    USUBJID = sprintf("SUB%04d", seq_len(n)),
    TRT01PN = trt,
    TRT01A = ifelse(trt == 1, "Experimental", "Control"),
    AVAL = round(ttev, 4),
    CNSR = 1 - ev,
    EVNT = ev,
    SEX = sex,
    ECOG = ecog,
    stringsAsFactors = FALSE
  )
}

if (data_csv != "") {
  adtte <- read.csv(data_csv, stringsAsFactors = FALSE)
} else {
  adtte <- generate_adtte(n, seed)
}

# --- KM estimation by treatment arm ---
library(survival)
library(jsonlite)

fit <- survfit(Surv(AVAL, EVNT) ~ TRT01PN, data = adtte)

# Extract curve at standard time points
time_points <- c(0, 3, 6, 9, 12, 15, 18, 21, 24, 30)

extract_curve <- function(fit, arm_code, time_points) {
  # Get the curve for this arm
  sd <- summary(fit, times = time_points, extend = TRUE)
  idx <- which(sd$strata == sprintf("TRT01PN=%d", arm_code))
  if (length(idx) == 0) {
    # Single stratum: no strata column
    surv_vals <- sd$surv
    lower <- sd$lower
    upper <- sd$upper
    n_risk <- sd$n.risk
    n_event <- sd$n.event
    n_censor <- sd$n.censor
  } else {
    surv_vals <- sd$surv[idx]
    lower <- sd$lower[idx]
    upper <- sd$upper[idx]
    n_risk <- sd$n.risk[idx]
    n_event <- sd$n.event[idx]
    n_censor <- sd$n.censor[idx]
  }
  list(
    time_points = as.numeric(time_points),
    survival = as.numeric(round(surv_vals, 6)),
    ci_lower = as.numeric(round(lower, 6)),
    ci_upper = as.numeric(round(upper, 6)),
    n_at_risk = as.integer(n_risk),
    n_events = as.integer(n_event),
    n_censored = as.integer(n_censor)
  )
}

curve_exp <- extract_curve(fit, 1, time_points)
curve_ctl <- extract_curve(fit, 0, time_points)

# Overall median
overall_fit <- survfit(Surv(AVAL, EVNT) ~ 1, data = adtte)
overall_median <- as.numeric(summary(overall_fit)$table["median"])
overall_ci <- as.numeric(summary(overall_fit)$table[c("0.95LCL", "0.95UCL")])

# Log-rank test
lr <- survdiff(Surv(AVAL, EVNT) ~ TRT01PN, data = adtte)
lr_chisq <- as.numeric(lr$chisq)
lr_df <- length(lr$n) - 1
lr_p <- as.numeric(1 - pchisq(lr_chisq, lr_df))

# Total subjects and events by arm
n_exp <- sum(adtte$TRT01PN == 1)
n_ctl <- sum(adtte$TRT01PN == 0)
events_exp <- sum(adtte$EVNT[adtte$TRT01PN == 1])
events_ctl <- sum(adtte$EVNT[adtte$TRT01PN == 0])

result <- list(
  tc_id = "TC-015",
  tc_name = "KM Curve with Risk Table",
  metadata = list(
    n_total = nrow(adtte),
    n_experimental = n_exp,
    n_control = n_ctl,
    events_experimental = events_exp,
    events_control = events_ctl,
    population = "ITT",
    time_unit = "months",
    time_points = time_points,
    r_version = R.version.string,
    package = "survival"
  ),
  overall_median = list(
    median = round(overall_median, 4),
    ci_lower = round(overall_ci[1], 4),
    ci_upper = round(overall_ci[2], 4)
  ),
  logrank = list(
    chisq = round(lr_chisq, 6),
    df = lr_df,
    p_value = round(lr_p, 6)
  ),
  curve_experimental = curve_exp,
  curve_control = curve_ctl
)

# Round metadata n's to integers
result$metadata$n_total <- as.integer(result$metadata$n_total)
result$metadata$n_experimental <- as.integer(result$metadata$n_experimental)
result$metadata$n_control <- as.integer(result$metadata$n_control)
result$metadata$events_experimental <- as.integer(result$metadata$events_experimental)
result$metadata$events_control <- as.integer(result$metadata$events_control)

json <- toJSON(result, auto_unbox = TRUE, pretty = TRUE)

if (output != "") {
  writeLines(json, output)
  cat(sprintf("Written: %s\n", output))
} else {
  cat(json, "\n")
}

File: benchmarks/references/ground-truth/Python/tc_015_km_curve.py
#!/usr/bin/env python3
"""TC-015: Kaplan-Meier Curve with Risk Table
Generates KM survival curve coordinates and risk table counts.
Output: JSON with curve coordinates, CI, and risk set at specified time points.

Dependencies: lifelines, numpy, pandas
"""

import argparse
import json
import sys
import numpy as np
import pandas as pd


def generate_adtte(n: int, seed: int) -> pd.DataFrame:
    """Generate synthetic ADTTE dataset matching R structure."""
    rng = np.random.default_rng(seed)
    n_exp = n // 2
    n_ctl = n - n_exp
    trt = np.array([1] * n_exp + [0] * n_ctl)
    # Shuffle treatment assignment
    rng.shuffle(trt)

    # Survival times: gamma distribution
    ttev = np.where(
        trt == 1,
        rng.gamma(shape=1.3, scale=14.0, size=n),
        rng.gamma(shape=1.1, scale=8.0, size=n),
    )
    # Censoring: uniform
    cens = rng.uniform(low=6, high=30, size=n)
    ev = (ttev <= cens).astype(int)
    ttev = np.minimum(ttev, cens)

    # Strata
    sex = rng.choice(["F", "M"], size=n)
    ecog = rng.choice([0, 1], size=n, p=[0.6, 0.4])

    return pd.DataFrame(
        {
            "USUBJID": [f"SUB{i:04d}" for i in range(1, n + 1)],
            "TRT01PN": trt.astype(int),
            "TRT01A": np.where(trt == 1, "Experimental", "Control"),
            "AVAL": np.round(ttev, 4),
            "CNSR": 1 - ev,
            "EVNT": ev,
            "SEX": sex,
            "ECOG": ecog,
        }
    )


def extract_curve(kmf, time_points, label):
    """Extract KM curve at specified time points using lifelines."""
    from lifelines import KaplanMeierFitter

    surv_at_tps = []
    ci_lower_at_tps = []
    ci_upper_at_tps = []
    n_at_risk = []
    n_events_at_tps = []
    n_censored_at_tps = []

    ci_df = kmf.confidence_interval_

    for tp in time_points:
        # Survival probability at time t
        s = kmf.predict(tp)
        surv_at_tps.append(round(float(s), 6))

        # CI at time t — lifelines CI is indexed by timeline
        # Get the closest timeline point <= tp
        mask = kmf.survival_function_.index <= tp
        if mask.any():
            closest_idx = kmf.survival_function_.index[mask][-1]
            ci_row = ci_df.loc[closest_idx]
            ci_lower_at_tps.append(round(float(ci_row.iloc[0]), 6))
            ci_upper_at_tps.append(round(float(ci_row.iloc[1]), 6))
        else:
            ci_lower_at_tps.append(1.0)
            ci_upper_at_tps.append(1.0)

    # Number at risk at each time point
    risk_tbl = kmf.event_table
    for tp in time_points:
        at_risk_mask = (risk_tbl.index <= tp)
        if at_risk_mask.any():
            idx = np.searchsorted(risk_tbl.index.values, tp, side="right") - 1
            if idx < 0:
                idx = 0
            n_at_risk.append(int(risk_tbl.iloc[idx]["at_risk"]))
            n_events_at_tps.append(int(risk_tbl.iloc[idx]["observed"]))
            n_censored_at_tps.append(int(risk_tbl.iloc[idx]["censored"]))
        else:
            n_at_risk.append(int(risk_tbl.iloc[0]["at_risk"]))
            n_events_at_tps.append(0)
            n_censored_at_tps.append(0)

    return {
        "time_points": [float(t) for t in time_points],
        "survival": surv_at_tps,
        "ci_lower": ci_lower_at_tps,
        "ci_upper": ci_upper_at_tps,
        "n_at_risk": n_at_risk,
        "n_events": n_events_at_tps,
        "n_censored": n_censored_at_tps,
    }


def main():
    parser = argparse.ArgumentParser(description="TC-015: KM Curve with Risk Table")
    parser.add_argument("--seed", type=int, default=42)
    parser.add_argument("--n", type=int, default=200)
    parser.add_argument("--output", type=str, default="")
    parser.add_argument("--data", type=str, default="")
    args = parser.parse_args()

    print(f"[{pd.Timestamp.now()}] TC-015 KM Curve + Risk Table | seed={args.seed} n={args.n}")

    if args.data:
        adtte = pd.read_csv(args.data)
    else:
        adtte = generate_adtte(args.n, args.seed)

    from lifelines import KaplanMeierFitter
    from lifelines.statistics import multivariate_logrank_test

    time_points = [0, 3, 6, 9, 12, 15, 18, 21, 24, 30]

    # Fit KM for each arm
    kmf_exp = KaplanMeierFitter()
    kmf_ctl = KaplanMeierFitter()

    exp_df = adtte[adtte["TRT01PN"] == 1]
    ctl_df = adtte[adtte["TRT01PN"] == 0]

    kmf_exp.fit(exp_df["AVAL"], event_observed=exp_df["EVNT"], label="Experimental")
    kmf_ctl.fit(ctl_df["AVAL"], event_observed=ctl_df["EVNT"], label="Control")

    curve_exp = extract_curve(kmf_exp, time_points, "Experimental")
    curve_ctl = extract_curve(kmf_ctl, time_points, "Control")

    # Overall median
    kmf_all = KaplanMeierFitter()
    kmf_all.fit(adtte["AVAL"], event_observed=adtte["EVNT"])
    from lifelines.utils import median_survival_times

    overall_median_val = median_survival_times(kmf_all)
    overall_ci = median_survival_times(kmf_all.confidence_interval_)

    # Log-rank test
    lr_result = multivariate_logrank_test(
        adtte["AVAL"], adtte["TRT01PN"], adtte["EVNT"]
    )
    lr_chisq = float(lr_result.test_statistic)
    lr_p = float(lr_result.p_value)

    result = {
        "tc_id": "TC-015",
        "tc_name": "KM Curve with Risk Table",
        "metadata": {
            "n_total": int(len(adtte)),
            "n_experimental": int(len(exp_df)),
            "n_control": int(len(ctl_df)),
            "events_experimental": int(exp_df["EVNT"].sum()),
            "events_control": int(ctl_df["EVNT"].sum()),
            "population": "ITT",
            "time_unit": "months",
            "time_points": time_points,
            "python_version": sys.version.split()[0],
            "package": "lifelines",
        },
        "overall_median": {
            "median": round(float(overall_median_val), 4),
            "ci_lower": round(float(overall_ci.iloc[0, 0]), 4),
            "ci_upper": round(float(overall_ci.iloc[0, 1]), 4),
        },
        "logrank": {
            "chisq": round(lr_chisq, 6),
            "df": 1,
            "p_value": round(lr_p, 6),
        },
        "curve_experimental": curve_exp,
        "curve_control": curve_ctl,
    }

    js = json.dumps(result, indent=2)
    if args.output:
        with open(args.output, "w") as f:
            f.write(js)
        print(f"Written: {args.output}")
    else:
        print(js)


if __name__ == "__main__":
    main()
